tensorrt_llm/serve/cluster_storage.py

```diff
@@ -0,0 +1,370 @@
+import abc
+import asyncio
+import logging
+import time
+from dataclasses import dataclass
+from enum import IntEnum
+from functools import wraps
+from typing import Dict, List, Optional
+
+import aiohttp
+from fastapi import FastAPI
+from fastapi.responses import JSONResponse
+from pydantic import BaseModel
+
+logger = logging.getLogger('uvicorn.error')
+
+
+class StorageItem(BaseModel):
+    key: str
+    value: Optional[str] = ""
+    expire_time: Optional[int] = -1
+    ttl: Optional[int] = -1
+    overwrite_if_exists: Optional[bool] = False
+
+
+class WatchEventType(IntEnum):
+    SET = 0
+    DELETE = 1
+
+
+@dataclass
+class WatchEvent:
+    storage_item: StorageItem
+    event_type: WatchEventType
+
+
+class WatchEventQueue:
+
+    def __init__(self, key_prefixes: List[str],
+                 events: asyncio.Queue[WatchEvent]):
+        self.key_prefixes = key_prefixes
+        self.events = events
+
+    async def drain(self):
+        events = []
+        event = await self.events.get()
+        logger.debug(f"Draining watch event: {self.events.qsize()}")
+        events.append(event)
+        while not self.events.empty():
+            event = self.events.get_nowait()
+            events.append(event)
+        self.events.task_done()
+        logger.debug(f"after draining watch event: {self.events.qsize()}")
+        return events
+
+
+class ClusterStorage(abc.ABC):
+
+    def __init__(self, cluster_uri: str, cluster_name: str):
+        ...
+
+    # start the storage, if it's already started, do nothing
+    async def start(self):
+        ...
+
+    # stop the storage, if it's already stopped, do nothing
+    async def stop(self):
+        ...
+
+    async def set(self,
+                  key: str,
+                  value: str,
+                  overwrite_if_exists=False,
+                  ttl: int = -1) -> bool:
+        ...
+
+    # refresh the key’s ttl
+    async def expire(self, key: str, ttl: int) -> bool:
+        ...
+
+    async def get(self, key: str) -> str:
+        ...
+
+    async def delete(self, key: str) -> bool:
+        ...
+
+    async def watch(self, key_prefix: str) -> WatchEventQueue:
+        ...
+
+    async def unwatch(self, key_prefix: str) -> None:
+        ...
+
+    async def get_prefix(self,
+                         key_prefix: str,
+                         keys_only: bool = False) -> Dict[str, str]:
+        ...
+
+
+def create_cluster_storage(cluster_uri, cluster_name, **kwargs):
+    if cluster_uri.startswith("http"):
+        return HttpClusterStorageServer(cluster_uri, cluster_name, **kwargs)
+    raise ValueError(f"Invalid cluster storage URI: {cluster_uri}")
+
+
+def create_cluster_storage_client(cluster_uri, cluster_name):
+    if cluster_uri.startswith("http"):
+        return HttpClusterStorageClient(cluster_uri, cluster_name)
+    raise ValueError(f"Invalid cluster storage URI: {cluster_uri}")
```

Comment:
Should we have a `ClusterStorageType` instead of using the uri to distinguish between the different storage types?